torch_glow/src/PyTorchModelLoader.cpp

```diff
@@ -25,47 +28,354 @@ glow::Type ptTypeToGlowType(const at::CompleteTensorType &ptType) {
+  glow::NodeValue lhs = getGlowNodeValue(inputs[1]);
+  glow::DivNode *glowNode = f_->createDiv("div", rhs, lhs);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
+
+void PyTorchModelLoader::loadAdd(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 3);
+  assert(outputs.size() == 1);
+
+  // TODO: extend this to allow non-constant scalars.
+  // Check scalar is 1, any other value is not supported.
+  const auto scalarHandle = getGlowConstantHandle<int32_t>(inputs[2]);
+  assert(scalarHandle.size() == 1);
+  assert(scalarHandle.raw(0) == 1);
+
+  glow::NodeValue rhs = getGlowNodeValue(inputs[0]);
+  glow::NodeValue lhs = getGlowNodeValue(inputs[1]);
+  glow::AddNode *glowNode = f_->createAdd("add", rhs, lhs);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
+
+void PyTorchModelLoader::loadSub(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 3);
+  assert(outputs.size() == 1);
+
+  // TODO: extend this to allow non-constant scalars.
+  // Check scalar is 1, any other value is not supported.
+  const auto scalarHandle = getGlowConstantHandle<int32_t>(inputs[2]);
+  assert(scalarHandle.size() == 1);
+  assert(scalarHandle.raw(0) == 1);
+
+  glow::NodeValue rhs = getGlowNodeValue(inputs[0]);
+  glow::NodeValue lhs = getGlowNodeValue(inputs[1]);
+  glow::SubNode *glowNode = f_->createSub("sub", rhs, lhs);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
+
+void PyTorchModelLoader::loadRelu(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 1);
+  assert(outputs.size() == 1);
+
+  glow::NodeValue input = getGlowNodeValue(inputs[0]);
+  glow::ReluNode *glowNode = f_->createRELU("relu", input);
+  addGlowNodeValue(outputs[0], glowNode->getResult());
+}
+
+void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 12);
+  assert(outputs.size() == 1);
+
+  // Glow expects conv inputs to be in NHWC but PyTorch keeps them in NCHW so we
+  // tranpose them.
+  glow::NodeValue input = getGlowNodeValue(inputs[0]);
+  input = f_->createTranspose("conv_input_transposed", input, NCHW2NHWC);
+  glow::ShapeNHWC inputShape(input.dims());
+
+  // Glow expects conv weights to be in CRSK but PyTorch keeps them in CKRS
+  // so we tranpose them. C - output_depth, R - filter_height, S -
+  // filter_width, K - input_depth.
+  glow::NodeValue weights = getGlowNodeValue(inputs[1]);
+  weights = f_->createTranspose("conv_weights_transposed", weights, NCHW2NHWC);
+  glow::ShapeNHWC weightsShape(weights.dims());
+
+  // If a bias was provided then use it otherwise create a 0 bias.
+  glow::NodeValue bias;
+  if (hasGlowNodeValue(inputs[2])) {
```

Comment:
One question: in jit if we dont have a param, when do we suppose to expect "hasGlowNodeValue" to be false, and when do we expect its value to be nonsense (like transposeHandle)?

Reply:
conv bias is optional, it is specified by a constant with None output if there is no bias, we handle this by loading None as nothing. This gives basically the same behavior as the c2 model loader.